### DISCUSSION TITLE
Tinyusb with i2c, spi, uart? using STM32.

### DISCUSSION THREAD

**Discussion by andolon** [hathach/tinyusb#1758] [Q&A] [answered]
A newbie's question:  How to integrate Tinyusb with Stm CubeMx ide? 
I need HAL features (**i2c, spi, uart**).
I followed this guide for integration: 
[https://anvilelectronics.ovh/tinyusb-na-stm32-device-cdc/](https://anvilelectronics.ovh/tinyusb-na-stm32-device-cdc/) 
I've used the makefiles from Tinyusb folder to compile.  Finally my project **works fine**.  
I`ve tried other features (i2c, usi, uart) with _TinyUsb own haldriver_  but nothing works.   
Example: Using Uart to send a char. Compiling and running with native HAL is ok, I can read chars with serial terminal. 
But including --> "mcu/st/stm32f1xx_hal_driver/Inc/stm32f1xx_hal_uart.h" and its .c doesn't send chars.
My board is STM32 f103c8t6 blue pill.  Thanks.

**Comment by Nikitarc** [comment, marked as answer]
Do not use HAL provided with  tinyusb. Put the HAL include path to your HAL files before the include of tinyusb.
In my development I entirely remove the tinyusb HAL directories...
